### ISSUE TITLE
Option to capture by `clone()` in `run!{}`?

### ISSUE THREAD

**Issue by soulsource** [bodil/higher#6]
# The problem
While working with the `run!{}` macro I noticed that it's problematic to work with types that aren't `Copy` (for instance `std::rc::Rc`). For example, the following code does not compile:
```Rust
#[derive(Clone)]
pub struct Test{}

pub fn test() -> Option<Test>{
    run!{
        t <= Some(Test{});
        Some(Test{});
        Some(Test{});
        Some(t.clone())
    }
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1ab8cc716dbfa9b2291b0e42416e7a6f))

## Explanation
This is a consequence of how `run!{}` desugars. The above code is effectiveley desugared to:
```Rust
#[derive(Clone)]
pub struct Test{}

pub fn test() -> Option<Test>{
    Some(Test{})
        .bind(move |t| Some(Test{})
            .bind(move |_| Some(Test{})
                .bind(move |_| Some(t.clone()))
            )
        )
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ee490db7732bffd4c3ab0fd3f45df66d))

Since `bind()` needs to be able to call its parameter several times (`Fn`, not `FnOnce`), this cannot compile, because the same value would need to be moved multiple times.

# Existing workarounds
## `Copy`
The obvious workaround would be to just make the data `Copy` ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fa15fa013d81ac52c428d3a2b319c1ae)).
This is by far the easiest solution, and as long as the data that is captured is just a hand full of bytes, that's probably the best way to do this anyhow.

It's quite often not possible to mark a type as `Copy` though.

## Leaving do-notation
One other option would be to leave do-notation, to bring the value into the enclosing scope, but that leads to quite convoluted syntax...
```Rust
#[derive(Clone)]
pub struct Test{}

pub fn test() -> Option<Test>{
    let t = Some(Test{});
    run! {
        t.clone();
        Some(Test{});
        Some(Test{})
    }
    .bind(move |_| t.clone().bind(Some))
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1973d8b6eb653a75912d6a08ddda074e))

## Passing the value through
This is basically the same as just marking the type as `Copy`, but more explicit, and also works for types that cannot be `Copy`.
```Rust
#[derive(Clone)]
pub struct Test{}

pub fn test() -> Option<Test>{
    run!{
        t <= Some(Test{});
        t <= Some(Test{}).fmap(move |_| t.clone());
        t <= Some(Test{}).fmap(move |_| t.clone());
        Some(t)
    }
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bc4050864ea307a703c1afe12f0970ae))

# Potential solution?
I don't think that there is a viable solution that completely avoids cloning, but having some shorter syntax than the one presented in the last workaround might be helpful.

A working desugared solution might look like this:
```Rust
#[derive(Clone)]
pub struct Test{}

pub fn test() -> Option<Test>{
    Some(Test{})
        .bind(|t| {let t = t.clone(); Some(Test{})
            .bind(|_| { let t = t.clone(); Some(Test{})
                .bind(|_| {let t = t.clone(); Some(t)})
            })
        })
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2739e8fb6e10986a8304a0c75ce317a1))

I would not generate the shadow clones implicitly though, because it is not clear when the chain of clones should stop. Also it wouldn't be very transparent to the user, regarding CPU costs.

From a user's perspective, I think it would be nice to have a way to specify a list of to-be-cloned parameters, similar to the already added `<$coerce:ident>` parameter:

```Rust
#[derive(Clone)]
pub struct Test{}

pub fn test() -> Option<Test>{
    run!{
        t <= Some(Test{});
        Some(Test{});
        [t] Some(Test{});
        [t] Some(t)
    }
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9592d5aa4e5ac72c34972d4748153c11))

# Summary
I think it might be worth to add a list of to-be-cloned identifiers to the `run!{}` macro. The primary use case would be to pass heap-objects like `std::rc::Rc` down the nested chain of binds, but it would also be useful for any other non-`Copy` types.

The main drawback would be added complexity, both for maintaining the `run!{}` macro, and for using it.

What do you think? Is this a good idea? Should I take the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9592d5aa4e5ac72c34972d4748153c11) example and make it a proper pull request, or is this an additional maintenance burden that should rather be avoided, given that there are workarounds?

**Comment by bodil** [comment]
This looks useful to me, I'd take a PR. I'm not sure I like the syntax, but I also can't currently come up with anything clearer, so let's just go with it?

**Comment by soulsource** [comment]
Yeah, I've been thinking about the syntax too, but I'm primed by C++, so I keep ending up with `[x,y]` again and again.

I also don't really like the further duplication of match arms in the macro, and I hope to find a non-ambiguous version that doesn't need it while working on the actual PR.

**Comment by bodil** [comment]
Closed by #7.
